packages/puppeteer-core/src/bidi/HTTPRequest.ts

```
@@ -5,107 +5,126 @@
     super();
+    requests.set(request, this);
 
-    this.#url = event.request.url;
-    this.#resourceType = event.initiator.type.toLowerCase() as ResourceType;
-    this.#method = event.request.method;
-    this.#postData = undefined;
-    this.#initiator = event.initiator;
+    this.#request = request;
     this.#frame = frame;
+    this.id = request.id;
+  }
 
-    this.id = event.request.request;
-    this._redirectChain = redirectChain;
-    this._navigationId = event.navigation;
-
-    for (const header of event.request.headers) {
-      // TODO: How to handle Binary Headers
-      // https://w3c.github.io/webdriver-bidi/#type-network-Header
-      if (header.value.type === 'string') {
-        this.#headers[header.name.toLowerCase()] = header.value.value;
-      }
-    }
+  override get client(): CDPSession {
+    throw new Error('Method not implemented.');
   }
 
-  override get client(): never {
-    throw new UnsupportedOperation();
+  #initialize() {
+    this.#request.on('redirect', request => {
+      this.#redirect = BidiHTTPRequest.from(request, this.#frame);
+    });
+    this.#request.once('success', data => {
+      this.#response = BidiHTTPResponse.from(data, this);
+    });
+
+    this.#frame?.page().emit(PageEvent.Request, this);
   }
 
   override url(): string {
-    return this.#url;
+    return this.#request.url;
   }
 
   override resourceType(): ResourceType {
-    return this.#resourceType;
+    return this.initiator().type.toLowerCase() as ResourceType;
   }
 
   override method(): string {
-    return this.#method;
+    return this.#request.method;
   }
 
   override postData(): string | undefined {
-    return this.#postData;
+    return undefined;
   }
 
   override hasPostData(): boolean {
-    return this.#postData !== undefined;
+    return false;
   }
 
   override async fetchPostData(): Promise<string | undefined> {
-    return this.#postData;
+    return undefined;
```

Let's throw UnsupportedOperation